### PR TITLE
Implementation of an algorithm to compute Chebyshev polynomials

### PR DESCRIPTION
**Pull request by haru-44**
Currently, `dup_chebyshevt` calculates $T_n$ in $O(n)$ using the recurrence relation $T_n = 2xT_{n-1} - T_{n-2}$. Additionally, we have implemented an algorithm that utilizes $T_{2n} = 2T_n^2 - 1$ and $T_{2n+1} = 2T_{n+1}T_n - x$ to compute in $O(\log n)$.

https://en.wikipedia.org/wiki/Chebyshev_polynomials

#### Release Notes

* polys
    * `chebyshevt_poly(n, x)` was made faster for large `n`.

### CODE REVIEW

File: sympy/polys/orthopolys.py [hunk 1]
```diff
@@ -72,10 +72,24 @@ def dup_chebyshevt(n, K):
...
+    # T_{2n} = 2T_n^2 - 1
+    # T_{2n+1} = 2T_{n+1}T_n - x
+    m2, m1 = [K.one, K.zero], [K(2), K.zero, -K.one] # T_1, T_2
+    for i in bin(n)[3:]:
+        c = dup_sub_term(dup_mul_ground(dup_mul(m1, m2, K), K(2), K), K.one, 1, K)
+        if  i  == '1':
+            m2, m1 = c, dup_sub_ground(dup_mul_ground(dup_sqr(m1, K), K(2), K), K.one, K)
+        else:
+            m2, m1 = dup_sub_ground(dup_mul_ground(dup_sqr(m2, K), K(2), K), K.one, K), c
+    return m2
```
Comment (oscarbenjamin) [inline sympy/polys/orthopolys.py:92]: It would be better to make separate functions for each algorithm and then have e.g.:
```
if n < 64:
    return dup_chebyshevt_small(...)
else:
    return dup_chebyshevt_large(...)
```
Perhaps better names can be chosen. Having separate functions is useful so that we can compare their outputs.

Comment (haru-44) [reply sympy/polys/orthopolys.py]: Created `_dup_chebyshevt_rec` and `_dup_chebyshevt_prod`. I am unsure about the appropriateness of the names.

File: sympy/polys/orthopolys.py [hunk 2]
```diff
@@ -72,11 +72,74 @@ def dup_chebyshevt(n, K):
...
+    ===========
+
+    Chebyshev polynomials of the first kind are defined by the recurrence relation:
+
+    .. math::
+        T_0(x) &= 1\\
+        T_1(x) &= x\\
+        T_n(x) &= 2xT_{n-1}(x) - T_{n-2}(x)
+
+    This function calculates the Chebyshev polynomial of the first kind using the above recurrence relation.
```
Comment (oscarbenjamin) [inline sympy/polys/orthopolys.py:93]: Long lines (especially in docstrings) should be wrapped to max length 79 characters. Many editors can do this for you e.g. I use vim and I can just select lines and use `gq` to reformat them.

File: sympy/polys/orthopolys.py [hunk 3]
```diff
@@ -72,11 +72,74 @@ def dup_chebyshevt(n, K):
     """Low-level implementation of Chebyshev polynomials of the first kind."""
     if n < 1:
         return [K.one]
+    # When n is small, it is faster to directly calculate the recurrence relation.
+    if n < 64: # The threshold serves as a heuristic
+        return _dup_chebyshevt_rec(n, K)
+    return _dup_chebyshevt_prod(n, K)
+
+def _dup_chebyshevt_rec(n, K):
+    r""" Low-level implementation of Chebyshev polynomials of the first kind using a recurrence relation.
```
Comment (oscarbenjamin) [inline sympy/polys/orthopolys.py:81]: I would just shorten this to `Chebyshev polynomials of the first kind using recurrence.`

File: sympy/polys/orthopolys.py [hunk 4]
```diff
@@ -72,11 +72,74 @@ def dup_chebyshevt(n, K):
...
+    is more efficient than using the standard recurrence relation for large n.
+
+    Parameters
+    ==========
+
+    n : int
+        n is a nonnegative integer.
+    K : domain
+
+    """
```
Comment (oscarbenjamin) [inline sympy/polys/orthopolys.py:133]: I don't think that these functions need such long docstrings especially if internal only (does this even appear in the html docs?).

I would shorten this to maybe
```python
def _dup_chebyshevt_prod(n, K):
    r"""Chebyshev T polynomials using recursive products

    Explanation
    ===========

    Computes Chebyshev T polynomials using

    .. math::
        T_{2n}(x) &= 2T_n^2(x) - 1\\
        T_{2n+1}(x) &= 2T_{n+1}(x)T_n(x) - x

    This is faster than ``_dup_chebyshevt_rec`` for large ``n``.

    Parameters
    ==========
    n : int (nonegative)
    K : domain
    """
```

**Comment by oscarbenjamin** [comment]
It isn't explicitly mentioned anywhere but I assume that you have tested this to see that it is faster for large `n` i.e. that is where the `n < 64` threshold comes from.

**Comment by oscarbenjamin** [comment]
This looks good to me. I would add a release note like:

* polys
    * `chebyshevt_poly(n, x)` was made faster for large `n`.

**Comment by haru-44** [comment]
> It isn't explicitly mentioned anywhere but I assume that you have tested this to see that it is faster for large `n` i.e. that is where the `n < 64` threshold comes from.

Yes.
I just experimented a bit, so there may be a better threshold.

File: sympy/polys/orthopolys.py (hunk 4 above)
Comment (haru-44) [reply sympy/polys/orthopolys.py]: I think it would be appreciated to have as much documentation as possible, even for functions for internal use.

Comment (oscarbenjamin) [reply sympy/polys/orthopolys.py]: I agree that documentation is good but "as much as possible" is not right: there has to be some limit. At some point it is better to understand what is happening by looking at the code than the documentation but some documentation of internals also helps with that. Right now functions in the polys module have too little documentation but it is definitely also possible to go too far in the other direction especially here when we have multiple functions that all take the same arguments and return the exact same thing.

Also we need to consider who this documentation is for i.e. who is going to read about the internal functions in this module and we can we reasonably expect them to know? Here I would say that the reader is either a SymPy developer or someone else who understands the internals enough to be able to use such things.

Here showing the recursion relation is good because someone can see that that is the basis of the correctness of the code. It is also the significant thing that is different about this function vs the other two functions that also return the same thing. Explaining that it is faster than the other function in some cases is good because that gives the background for the purpose of the code and again contrasts it with the two other functions.

Otherwise I don't think that (in this particular case) more is needed. If someone wants to measure/analyse the performance/complexity then they can but we should only mention here the main things needed to understood purpose/correctness or something that is very not obvious.

In this particular case also the comment about `O(log(n))` computations is misleading since it implicitly compares one method that uses only addition and scalar multiplication with another that uses polynomial multiplication (a more expensive operation).

**Comment by oscarbenjamin** [comment]
Looks good!

**Comment by haru-44** [comment]
thanks
